File: codes/2023/25-WildcardExcellence-Beihang (Java)/src/midend/value/global/GlobalVariable.java
package midend.value.global;

import midend.MyModule;
import midend.Type;
import midend.value.Value;

public class GlobalVariable extends Value {
    private Boolean isConst;
    private Value initialValue;

    public Type getInitialType() {
        return initialType;
    }

    private Type initialType;

    public GlobalVariable(Type type, String name, Value value, boolean isConst, Type initialType) {
        //globalVariable的 type 也为 pointerType, name 为 %g
        //全局变量未初始化时初始化为 0
        super(type, name);
        setType(new Type.PointerType(type));
        setName("@global" + GLO_NUM++);
        this.isConst = isConst;
        this.initialValue = value;
        this.initialType = initialType;
        MyModule.myModule.addGlobalVariable(this);
    }

    public Value getInitialValue() {
        return this.initialValue;
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(getName()).append(" = ");
        if (isConst) {
            sb.append("constant ");
        } else {
            sb.append("global ");
        }
        sb.append((getType()).getElementType()).append(" ");
        sb.append(this.initialValue.getName());
        return sb.toString();
    }
}

File: codes/2025/24-Excellence-USTC-100-19-37 (C++)/src/ir/GlobalVariable.cpp
#include "GlobalVariable.hpp"
#include "Module.hpp"

GlobalVariable::GlobalVariable(std::string name, Module *m, Type *ty,
                               bool is_const, Constant *init)
    : User(ty, name), is_const_(is_const), init_val_(init) {
    m->add_global_variable(this);
    if (init) {
        this->add_operand(init);
    }
} // global操作数为initval

GlobalVariable *GlobalVariable::create(std::string name, Module *m, Type *ty,
                                       bool is_const,
                                       Constant *init = nullptr) {
    return new GlobalVariable(name, m, PointerType::get(ty), is_const, init);
}